drain/drain.py

```diff
@@ -0,0 +1,417 @@
+
+        fees = route[0]['amount_msat'] - route[-1]['amount_msat']
+
+        # check fee and exclude worst channel the next time
+        # NOTE: the int(msat) casts are just a workaround for outdated pylightning versions
+        if fees > payload['exemptfee'] and int(fees) > int(payload['amount']) * payload['maxfeepercent'] / 100:
+            worst_channel_id = find_worst_channel(route)
+            if worst_channel_id is None:
+                raise RpcError(payload['command'], payload, {'message': 'Insufficient fee'})
+            excludes += [worst_channel_id + '/0', worst_channel_id + '/1']
+            continue
+
+        plugin.log("Sending over %d hops to %s %s using %s fees" % (len(route), payload['command'], payload['amount'], fees))
+        for r in route:
+            plugin.log("    - %s  %14s  %s" % (r['id'], r['channel'], r['amount_msat']))
+
+        try:
+            plugin.rpc.sendpay(route, payment_hash, label)
+            result = plugin.rpc.waitsendpay(payment_hash, payload['retry_for'] + start_ts - int(time.time()))
+            if result.get('status') == 'complete':
+                payload['success_msg'] += ["%dmsat sent over %d hops to %s %dmsat [%d/%d]" % (payload['amount'] + fees, len(route), payload['command'], payload['amount'], chunk+1, payload['chunks'])]
+                # we need to wait for gossipd to update to new state,
+                # so remaining amounts will be calculated correctly for the next chunk
+                spendable, _ = spendable_from_scid(plugin, payload, payload['scid'])
+                while spendable == spendable_before:
+                    time.sleep(0.5)
+                    spendable, _ = spendable_from_scid(plugin, payload, payload['scid'])
+                return True
+            return False
+
+        except RpcError as e:
+            erring_message = e.error.get('message', '')
+            erring_channel = e.error.get('data', {}).get('erring_channel')
+            erring_index = e.error.get('data', {}).get('erring_index')
+            erring_direction = e.error.get('data', {}).get('erring_direction')
+
+            # detect exceeding of HTLC commitment fee
+            if 'Capacity exceeded' in erring_message and erring_index == 0:
+                match = HTLC_FEE_PAT.search(erring_message);
+                if match:  # new servers tell htlc_fee via exception (#2691)
+                    raise ValueError("htlc_fee is %s" % match.group(1))
+                raise ValueError("htlc_fee unknown")
+
+            if erring_channel == payload['scid']:
+                raise RpcError(payload['command'], payload, {'message': 'Error with selected channel: %s' % erring_message})
+
+            plugin.log("RpcError: " + str(e))
+            if erring_channel is not None and erring_direction is not None:
+                excludes.append(erring_channel + '/' + str(erring_direction))
+
+
+def execute(command: str, scid: str, percentage: int,
```

Comment:
Trying this for real, I think it could be great if it could be a float as you may often need to not drain / fill a significant percentage of the channel funds

Reply:
thx, addressed by 26e8717